<a href="https://colab.research.google.com/github/puttipongyy/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import os

In [9]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [11]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [12]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

Quiz: ข้อมูลมาจากกี่ประเทศ

In [13]:
len(set(data['Country']))

38

In [14]:
data['Country'].unique()

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

# แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย *groupby.apply(list)*

In [15]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [16]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [17]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=0139dbfbd800f7d1a292c9adf8924fdd6dafd7358da99578a3448dddce1cf230
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [18]:
from apyori import apriori

In [19]:
apriori(transactions,min_support = 0.5)

<generator object apriori at 0x7e10a57fee30>

In [20]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [21]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [22]:
len(frequentItemsets)

4

In [23]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [24]:
frequentItemsets[0][0]

frozenset({'22423'})

In [25]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [26]:
frequentItemsets[0][0].__len__()

1

In [27]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [28]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

Quiz <br>
stock code 22386 and 85099B คือสินค้าอะไร

In [29]:
set(data[(data['StockCode']=='22386')]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [30]:
set(data[(data['StockCode']=='85099B')]['Description'])

{'JUMBO BAG RED RETROSPOT'}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'562950',
 '539564',
 '548648',
 '538903',
 'C538124',
 '555277',
 '580673',
 '566577',
 '552715',
 '549592',
 '570872',
 '541424',
 '565724',
 '567378',
 '543188',
 '541592',
 '549524',
 '554483',
 '566396',
 '538951',
 '576381',
 '547096',
 '572969',
 '579518',
 '555510',
 '552663',
 '557493',
 '572552',
 '550899',
 '577039',
 '543251',
 '546669',
 '555275',
 '551825',
 '567507',
 '578067',
 '549997',
 'C545438',
 '581135',
 '551994',
 '550986',
 '551058',
 '567886',
 '553542',
 '564184',
 '542583',
 '558119',
 '566482',
 '543099',
 'C566167',
 '550613',
 '573020',
 '549360',
 '564092',
 '545667',
 '549435',
 '576690',
 'C543633',
 '565784',
 '542862',
 '571746',
 '553489',
 '580053',
 '547378',
 '544913',
 '563561',
 '552247',
 '541695',
 '542600',
 '547830',
 '536544',
 '551173',
 '550437',
 '555400',
 '573822',
 '578031',
 '545377',
 'C556635',
 'C562921',
 '538148',
 '537240',
 '548733',
 '556083',
 'C560588',
 '566400',
 '561651',
 '551885',
 '556577',
 'C556880',
 '573892',
 '

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

#HW 1 หา Frequent Item Set และ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia  (15 Jan - 15 Feb)
* ของช่วงเวลา  Christmas & New Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* โดยจัดกลุ่มตามประเทศ + เดือน


In [31]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom

In [33]:

# Given data
countries = ['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
             'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
             'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
             'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
             'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
             'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
             'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
             'European Community', 'Malta', 'RSA']

# Fictional mapping of countries to continents (replace with accurate data)
continent_mapping = {
    'United Kingdom': 'Europe',
    'France': 'Europe',
    'Australia': 'Oceania',
    'Netherlands': 'Europe',
    'Germany': 'Europe',
    'Norway': 'Europe',
    'EIRE': 'Europe',
    'Switzerland': 'Europe',
    'Spain': 'Europe',
    'Poland': 'Europe',
    'Portugal': 'Europe',
    'Italy': 'Europe',
    'Belgium': 'Europe',
    'Lithuania': 'Europe',
    'Japan': 'Asia',
    'Iceland': 'Europe',
    'Channel Islands': 'Europe',
    'Denmark': 'Europe',
    'Cyprus': 'Asia',
    'Sweden': 'Europe',
    'Austria': 'Europe',
    'Israel': 'Asia',
    'Finland': 'Europe',
    'Bahrain': 'Asia',
    'Greece': 'Europe',
    'Hong Kong': 'Asia',
    'Singapore': 'Asia',
    'Lebanon': 'Asia',
    'United Arab Emirates': 'Asia',
    'Saudi Arabia': 'Asia',
    'Czech Republic': 'Europe',
    'Canada': 'North America',
    'Unspecified': 'Unknown',
    'Brazil': 'South America',
    'USA': 'North America',
    'European Community': 'Europe',
    'Malta': 'Europe',
    'RSA': 'Africa'
}

# Create a DataFrame
data_continent = {'Country': countries, 'Continent': [continent_mapping.get(country, 'Unknown') for country in countries]}
df = pd.DataFrame(data_continent)

# Display the DataFrame
df.head()


Country Continent
0  United Kingdom    Europe
1          France    Europe
2       Australia   Oceania
3     Netherlands    Europe
4         Germany    Europe

In [34]:
df

Country      Continent
0         United Kingdom         Europe
1                 France         Europe
2              Australia        Oceania
3            Netherlands         Europe
4                Germany         Europe
5                 Norway         Europe
6                   EIRE         Europe
7            Switzerland         Europe
8                  Spain         Europe
9                 Poland         Europe
10              Portugal         Europe
11                 Italy         Europe
12               Belgium         Europe
13             Lithuania         Europe
14                 Japan           Asia
15               Iceland         Europe
16       Channel Islands         Europe
17               Denmark         Europe
18                Cyprus           Asia
19                Sweden         Europe
20               Austria         Europe
21                Israel           Asia
22               Finland         Europe
23               Bahrain           Asia
24                Greece         Europe
25             Hong Kong           Asia
26             Singapore           Asia
27               Lebanon           Asia
28  United Arab Emirates           Asia
29          Saudi Arabia           Asia
30        Czech Republic         Europe
31                Canada  North America
32           Unspecified        Unknown
33                Brazil  South America
34                   USA  North America
35    European Community         Europe
36                 Malta         Europe
37                   RSA         Africa

In [36]:
df['Continent'].value_counts()

Europe           23
Asia              9
North America     2
Oceania           1
Unknown           1
South America     1
Africa            1
Name: Continent, dtype: int64

# เช็คประเทศ

In [37]:
df.sort_values(by = 'Continent')

Country      Continent
37                   RSA         Africa
29          Saudi Arabia           Asia
28  United Arab Emirates           Asia
27               Lebanon           Asia
26             Singapore           Asia
25             Hong Kong           Asia
23               Bahrain           Asia
21                Israel           Asia
14                 Japan           Asia
18                Cyprus           Asia
10              Portugal         Europe
35    European Community         Europe
30        Czech Republic         Europe
1                 France         Europe
3            Netherlands         Europe
4                Germany         Europe
5                 Norway         Europe
24                Greece         Europe
9                 Poland         Europe
22               Finland         Europe
7            Switzerland         Europe
6                   EIRE         Europe
19                Sweden         Europe
36                 Malta         Europe
17               Denmark         Europe
16       Channel Islands         Europe
15               Iceland         Europe
8                  Spain         Europe
13             Lithuania         Europe
12               Belgium         Europe
11                 Italy         Europe
20               Austria         Europe
0         United Kingdom         Europe
31                Canada  North America
34                   USA  North America
2              Australia        Oceania
33                Brazil  South America
32           Unspecified        Unknown

In [ ]:
table_merge = result_constituencies_Candidate.merge(Candidate_constituency, how ='left', left_on='mp_app_id',right_on='mp_app_id')
print(table_merge.shape)
table_merge.head()